### ISSUE TITLE
[4.1 unstable RPI3] Xiaomi Mi Gamepad bluetooth not detected

### ISSUE THREAD

**Issue by bigboo3000** [recalbox/recalbox-os#1220]
this gamepad is not detected at all in bluetooth (either when the pad is in normal mode or in pairing mode).

**Comment by substring** [comment]
Asked someone who owns one, it worked. Try forgetting known devices and pair again.

Anyway there is a forum thread for 4.1 bugs, better continue there

**Comment by bigboo3000** [comment]
I just tested with a fresh install of today build, I confirm you that my Mi gamepad doesn't appear in bluetooth list (I can see other bt pads). I rebooted and tried many times.

I'm using internal bluetooth of rpi3.

**Comment by pillabichos** [comment]
I have the same problem when updates to 4.1.
It's a Raspberry Pi 2

**Comment by substring** [comment]
Try by SSH `bluetoothctl`+ `scan on` and tell me if it gets listed

**Comment by idarius** [comment]
i can successfully pair my xiaomi gamepad.
Only problem is when the controller deconnect in game, i can't repair it on the fly, need to reboot the pi or kill / start ES to reconnect.

**Comment by bigboo3000** [comment]
ok the pad appears in SSH:

# bluetoothctl
[NEW] Controller B8:27:EB:25:E9:60 BlueZ 5.39 [default]
[NEW] Device AC:FD:93:CB:8F:1C 小米蓝牙手柄

Then I typed "scan on", but nothing appears in ES, maybe it's a problem with chinese caracters in the name?

**Comment by bigboo3000** [comment]
[CHG] Device AC:FD:93:CB:8F:1C RSSI: -73
[bluetooth]#
[CHG] Device D1:7B:D1:A3:4D:B1 RSSI: -51
[CHG] Device D1:7B:D1:A3:4D:B1 RSSI: -73
[bluetooth]# [NEW] Controller B8:27:EB:25:E9:60 BlueZ 5.39 [default]
Invalid command
[bluetooth]# [NEW] Device AC:FD:93:CB:8F:1C  Controller
Invalid command
[CHG] Device 5C:F6:DC:BD:77:EB RSSI: -97
[CHG] Device 5C:F6:DC:BD:77:EB RSSI: -87
[CHG] Device AC:FD:93:CB:8F:1C RSSI: -81
[CHG] Device AC:FD:93:CB:8F:1C RSSI: -61
[CHG] Device AC:FD:93:CB:8F:1C RSSI: -70
[CHG] Device 5C:F6:DC:BD:77:EB RSSI: -99
[bluetooth]#

**Comment by MaFL2k** [comment]
Any news on this topic? I have the same problem.

bluetoothctl+scan on shows the pad but it´s not listed in emulationstation

**Comment by bhayer** [comment]
exact same problem here.

shown by bluetoothctl, but not in emulationstation.

**Comment by cpcbegin** [comment]
I've got the same problem.
Please reopen this issue because people that we have this gamepad have this problem and we are a lot of people.

**Comment by turyvc** [comment]
Same problem, don't work with last 4.1. It ran on 4.0.2 and if I update to 4.1 it stop working( it don't find bluetooth). With other bluetooth controllers work, so there is a compatibility problem.

**Comment by substring** [comment]
I'm afraid the scanning problem can't be solved, it depends on bluez5 (remember there is a major leap in bluetooth management from recalbox 4.0 to 4.1). If you manually pair through bluetoothctl, don't forget to trust.

@Laddyacris also has problems with a different pad to reboot on connection, and it is solved with a restart of the bluetooth service `/etc/init.d/S32bluetooth restart`

Please test and report pairing by those 2 methods as I don't own such controller:
- `/recalbox/scripts/bluetooth/simple-agent.py AA:BB:CC:DD:EE:FF` with your right mac address in uppercase
- the full bluetoothctl version :
```
agent on
default-agent
power on
discoverable on
pairable on
scan on
pair AA:BB:CC:DD:EE:FF
trust AA:BB:CC:DD:EE:FF
```
And test that you can reconnect on reboot. Please note the pad must be turned on **AFTER** ES launched

**Comment by MaFL2k** [comment]
I had success with both methods.

-pairing with long keypress on the button (fast blinking)
-connecting after power on and reconnect after pad sleep with a short keypress (slow blinking)

Persistent.

````
[bluetooth]# paired-devices
[bluetooth]# agent on
Agent registered
[bluetooth]# default-agent
Default agent request successful
[bluetooth]# power on
Changing power on succeeded
[bluetooth]# discoverable on
Changing discoverable on succeeded
[bluetooth]# pairable on
Changing pairable on succeeded
[bluetooth]# scan on
Discovery started
[NEW] Device 1C:96:5A:DD:9F:5D 小米蓝牙手柄
[bluetooth]# pair 1C:96:5A:DD:9F:5D
Attempting to pair with 1C:96:5A:DD:9F:5D
[CHG] Device 1C:96:5A:DD:9F:5D Connected: yes
[CHG] Device 1C:96:5A:DD:9F:5D Modalias: usb:v2717p3144d0129
[CHG] Device 1C:96:5A:DD:9F:5D UUIDs: 00001000-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:96:5A:DD:9F:5D UUIDs: 00001124-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:96:5A:DD:9F:5D UUIDs: 00001200-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:96:5A:DD:9F:5D Paired: yes
Pairing successful
[CHG] Device 1C:96:5A:DD:9F:5D Connected: no
[bluetooth]# trust 1C:96:5A:DD:9F:5D
[CHG] Device 1C:96:5A:DD:9F:5D Trusted: yes
Changing 1C:96:5A:DD:9F:5D trust succeeded
[CHG] Device 1C:96:5A:DD:9F:5D RSSI: -75
[bluetooth]# connect 1C:96:5A:DD:9F:5D
Attempting to connect to 1C:96:5A:DD:9F:5D
[CHG] Device 1C:96:5A:DD:9F:5D Connected: yes
Connection successful
[小米蓝牙手柄]# paired-devices
Device 1C:96:5A:DD:9F:5D 小米蓝牙手柄
[CHG] Device 1C:96:5A:DD:9F:5D Connected: no
[CHG] Device 1C:96:5A:DD:9F:5D Class: 0x000508
[CHG] Device 1C:96:5A:DD:9F:5D Icon: input-gaming
[CHG] Device 1C:96:5A:DD:9F:5D Connected: yes
[CHG] Device D4:89:4A:E9:B5:AA RSSI: -77
[小米蓝牙手柄]#
````

````
# /recalbox/scripts/bluetooth/simple-agent 1C:96:5A:DD:96:5A
Agent registered
Device paired
#
````

**Comment by Nviard** [comment]
Hello, I've just installed recalbox and I have the same problem, but i don't understand your answers.

I tried to run bluetoothctl, but i get this error : bluetoothctl: command not found
How can I install it ?

Moreover, /recalbox/scripts/bluetooth/simple-agent.py  doesn't exist...

Can you help me please ?

**Comment by substring** [comment]
@Nviard you are not tunning recalbox 4.1, but 4.0.x, so this issue is not appropriate for you

**Comment by Nviard** [comment]
@substring You're right I'm sorry

PS : Xiaomi Gamepad is not detected on 4.0.1 either

**Comment by turyvc** [comment]
@Nviard Xiaomi Gamepad is detected on 4.0.1.
You should set it on pair mode(pulse central button for 5-10 sec)

But it is not detected on 4.1 unstable.
I will try it with the script.

**Comment by Edu-Murad** [comment]
Hello guys.
RPi3, Recalbox 4.1.0 unstable latest April release. Using Internal Pi BT.
With all controllers suport disable the USB controllers work fine, but when I enable PS3 controllers, no matter if I choose bluez5, or shanwan (mine is a ps3 clone) the emulationstation stops responding controllers command, no matter if USB or BT.
BT pairing works fine, the BT controller works for the first 5 secs, and then ES stops responding. BG music continues, and the upper right lightning blinks.

I have a 4.1.0 unstable march release in other SD, where the controllers are working fine, with some delay, but working.

Any suggestions?

**Comment by Laddyacris** [comment]
Hello
try this tuto https://github.com/recalbox/recalbox-os/wiki/Connect-bluetooth-controller-manually-%28EN%29

**Comment by Edu-Murad** [comment]
The controller is pairing, it works for some secs, after that my rpi3 stops responding, not freezing, just not responding for the usb controllers and bt controllers.

Anyway, I will try to do this procedure, reseting the controller and forgeting all paired devices.

**Comment by substring** [comment]
OK, had an opportunity to log on a recalbox from someone owning this pad, news aren't good ... Here is what I know so far :
1. recalbox doesn't list it in ES because the pad name uses some non standard caracters though crashing the software used to enumerate devices
2. might be dued to the pad being bluetooth 3.0, but it won't connect to recalbox by itself (which is bad)

So, here is what i've come up to in order to resolve this darn pad :
1. you have to mannually connect it using [this link](https://github.com/recalbox/recalbox-os/wiki/Connect-bluetooth-controller-manually-%28EN%29). Use 0000 as the pin code, don't forget to set the device in pairing mode
2. you have to edit `/etc/init.d/S32bluetooth` and replace `test-discovery` by `btDaemon` (there should be 3 occurences). Don't forget to `mount -o remount, rw /` first
3. reboot Recalbox. Once recalbox has rebooted, turn on your pad in pairing mode, it should auto connect

Please report if it worked or not :)

**Comment by bigboo3000** [comment]
If the problem is the chinese characters in the name, maybe this fix to retropie would be applicable to recalbox?

https://retropie.org.uk/forum/topic/4804/xiaomi-bluetooth-gamepad-after-4-0-6-update/4

post n°4:

In my quest to resolve the xiaomi bt controller issue, I also realised the configure BT script in retropie setup did not cater for utf 8 characters and resulted in the xiaomi controller not being able to show up in the "Display Registered & Connected Bluetooth Devices"

Did some research and added these lines in this script.
RetroPie-Setup/scriptmodules/supplementary/bluetooth/bluez-test-device

append under import sys
[code]
import sys
reload(sys)
sys.setdefaultencoding('utf-8')
[/code]

now my xiaomi bt gamepad shows up in the display bluetooth devices page instead of showing me this error msg

"unicodeencodeerror 'ascii' codec can't encode characters in position 18-23"

May I suggest this script to be updated to support BT devices with UTF-8 names?

**Comment by substring** [comment]
@bigboo3000 thanks for investigating ! Sadly this didn't work. The trick was to call the script with `PYTHONIOENCODING=UTF-8` and recalbox can then list the pad

**Comment by bigboo3000** [comment]
thank you for finding the solution! How can I test it?

**Comment by substring** [comment]
start with `mount -o remount, rw /` then edit `/recalbox/scripts/recalbox-configure.sh`. Find the line with `/recalbox/scripts/bluetooth/test-device list` and add the variable so that it looks like `PYTHONIOENCODING=UTF-8 /recalbox/scripts/bluetooth/test-device list`

You still need to do the previous trick as this pad won't auto connect to recalbox for an unknown reason (probable because it's bluetooth 3.0)

**Comment by bigboo3000** [comment]
Thank you I'll test that!

**Comment by cpcbegin** [comment]
I try change
`/recalbox/scripts/bluetooth/test-device list`
as 
`PYTHONIOENCODING=UTF-8 /recalbox/scripts/bluetooth/test-device list`
 with /recalbox/scripts/recalbox-config.sh (because I didn't found /recalbox/scripts/recalbox-configure.sh) but I still can't pair it, the menu says that I paired but it's not true.

I only can pair it follow all the steps explained in may, 5th but I need repeat all of them if the controller or raspberry is disconnected or  I don't use the controller for a while.

**Comment by substring** [comment]
@cpcbegin please read my previous comment : I never said the autoconnect works out of the box. You must do what i explained in https://github.com/recalbox/recalbox-os/issues/1220#issuecomment-315129245

thanks for correcting the script name, i'll edit my post

**Comment by cpcbegin** [comment]
I see the file /etc/init.d/S32bluetooth, Must I change only test-discovery by btDaemon or also test-discovery.pid by btDaemon.pid?

**Comment by substring** [comment]
all occurences !

**Comment by cpcbegin** [comment]
Eureka! now it I restart the raspberry pi, long press on xiaomi button and I can run a game with my gamepad.

Thanks for your efforts!

**Comment by substring** [comment]
solved with https://github.com/recalbox/recalbox-os/releases/tag/4.1.0
